lib/MySQL_HostGroups_Manager.cpp

```diff
@@ -4499,6 +4569,203 @@ void MySQL_HostGroups_Manager::read_only_action(char *hostname, int port, int re
 	free(query);
 }
 
+/**
+ * @brief New implementation of the read_only_action method that does not depend on the admin table.
+ *   The method checks each server in the provided list and adjusts the servers according to their corresponding read_only value.
+ *   If any change has occured, checksum is calculated.
+ *
+ * @param mysql_servers List of servers having hostname, port and read only value.
+ * 
+ */
+void MySQL_HostGroups_Manager::read_only_action_v2(const std::list<std::tuple<std::string,int,int>>& mysql_servers) {
+
+	std::string hostname;
+	int port = -1;
+	int read_only = -1;
+	bool update_mysql_servers_table = false;
+
+	unsigned long long curtime1 = monotonic_time();
+	wrlock();
+	for (const auto& server : mysql_servers) {
+		bool is_writer = false;
+		std::tie(hostname, port, read_only) = server;
+		const std::string& srv_id = hostname + ":::" + std::to_string(port);
+		
+		auto itr = hostgroup_server_mapping.find(srv_id);
+
+		if (itr == hostgroup_server_mapping.end()) {
+			proxy_warning("Server %s:%d not found\n", hostname.c_str(), port);
+			continue;
+		}
+
+		HostGroup_Server_Mapping* host_server_mapping = itr->second.get();
+
+		if (!host_server_mapping)
+			assert(0);
+
+		const std::vector<HostGroup_Server_Mapping::Node>& writer_map = host_server_mapping->get(HostGroup_Server_Mapping::Type::WRITER);
+
+		is_writer = !writer_map.empty();
+
+		if (read_only == 0) {
+			if (is_writer == false) {
+				// the server has read_only=0 (writer), but we can't find any writer, 
```

Comment:
I think we could benefit from a logging message, in every possible action that updates `update_mysql_servers_table`. That way we could tell in the logs which was the action responsible for the table 're-generation'. I'm going to try to propose a sensible one in each of the locations. For instance, for this one:

```
proxy_info("Server '%s:%d' found with 'read_only=0' but not found as writer\n", hostname.c_str(), port);
```

I think that for all the messages I'm going to propose, it could be OK to duplicate them also as debug messages:

```
proxy_debug(PROXY_DEBUG_MONITOR, 5, "Server '%s:%d' found with 'read_only=0' but not found as writer\n", hostname.c_str(), port);
```

I'm saying this thinking on other feature being developed... But we might want a better way of doing this.

Reply:
Done